common/ferrostar/src/navigation_controller/mod.rs

```diff
@@ -120,18 +120,31 @@ impl NavigationController {
                         let mut remaining_steps = remaining_steps.clone();
                         remaining_steps.remove(0);
 
+                        // we need to update the geometry index, since the step has changed
+                        let (updated_current_step_geometry_index, updated_snapped_user_location) =
+                            if let Some(current_route_step) = remaining_steps.first() {
+                                // TODO: We could move this to the Route struct or NavigationController directly to only calculate it once.
+                                let current_step_linestring = current_route_step.get_linestring();
+                                self.snap_user_to_line(
```

Comment:
should we be passing the original location here from the caller and using that instead of the snapped location?

Reply:
I also found this to be a struggle to understand as I built the snapping task.

@ianthetechie the core state logic feels like an area of the core where some substantial commenting may help us all down the road. Basically to:

1. Explain what params produce the major state changes like step advancement. 
2. To outline cases where unexpected params/vars are used (e.g. the different location vars).

Reply:
Yeah, agreed.

Reply:
I probably won't get around to commenting up this before my week ends, but I haven't forgotten about it. Alternately if @ahmedre ends up putting more time into this and identifies missing comments, adding them to this PR would be great too. Otherwise I can tackle improving docs here next week.